web/modules/custom/asset_distribution/asset_distribution.module

```
@@ -284,35 +284,44 @@ QUERY;
+ * @return \Drupal\rdf_entity\Entity\RdfInterface|null
+ *   The solution or release that this distribution belongs to.
  *
  * @throws \Exception
- *   Thrown when anything but an asset distribution entity is passed in.
+ *    Thrown when more than one parents are found.
  */
-function asset_distribution_title_is_unique(RdfInterface $asset_distribution) {
-  if ($asset_distribution->bundle() !== 'asset_distribution') {
-    throw new \Exception('Title uniqueness within a release can only be checked for asset distributions.');
+function asset_distribution_get_distribution_parent(RdfInterface $asset_distribution) {
+  if (!$asset_distribution->isNew()) {
+    $query = \Drupal::entityQuery('rdf_entity', 'OR')
+      ->condition('field_is_distribution', $asset_distribution->id())
+      ->condition('field_isr_distribution', $asset_distribution->id());
+    $parents = $query->execute();
+    if (count($parents) > 1) {
+      throw new Exception("More than one parents were found for distribution {$asset_distribution->label()}");
+    }
+    elseif (!empty($parents)) {
+      $rdf_entity_id = reset($parents);
+      return \Drupal::entityTypeManager()->getStorage('rdf_entity')->load($rdf_entity_id);
+    }
   }
-
-  $query = \Drupal::entityQuery('rdf_entity')
-    ->condition('label', $asset_distribution->label())
-    ->condition('rid', 'asset_distribution')
-    ->condition('og_audience', $asset_distribution->get('og_audience')->first()->getValue()['target_id']);
-  if (!empty($asset_distribution->id())) {
-    $query->condition('id', $asset_distribution->id(), '<>');
+  // If the distribution is now added, get the parent from the route.
+  elseif (\Drupal::routeMatch()->getRouteName() === 'asset_distribution.asset_release_asset_distribution.add') {
+    $parent = \Drupal::routeMatch()->getParameter('rdf_entity');
+    if ($parent instanceof RdfInterface && in_array($parent->bundle(), ['solution', 'asset_release'])) {
+      return $parent;
+    }
```

I had a look at how this can be fixed properly:

1. We can put a computed field on the distribution that contains the reference to the parent, similar to the one we put on solutions. It will get the value from the database for existing entities, and from a property on the entity object for new entities.
2. An `#entity_builder` firing in the form submission can populate the value on the computed field. This will cover the standard case of creating a distribution through the UI.
3. Other ways of creating (through federation or when running tests) can populate the value in a similar way, e.g. by setting a property on the entity that is used by the computed field.

This seems like a lot of work to block this ticket on. I will let this pass for now. Maybe we can throw an exception at this point if the value is not present on the route? That will alert us immediately if we are using this incorrectly.